File: Android/app/src/main/java/com/dreamproject/inwords/presentation/viewScenario/authorisation/AuthorisationViewState.java
package com.dreamproject.inwords.presentation.viewScenario.authorisation;

class AuthorisationViewState {
    public enum Status {
        LOADING, SUCCESS, ERROR
    }

    private static AuthorisationViewState SUCCESS_STATE = new AuthorisationViewState(Status.SUCCESS, null);
    private static AuthorisationViewState LOADING_STATE = new AuthorisationViewState(Status.LOADING, null);

    final Status status;
    final Throwable throwable;

    private AuthorisationViewState(Status status, Throwable throwable) {
        this.status = status;
        this.throwable = throwable;
    }

    static AuthorisationViewState success() {
        return SUCCESS_STATE;
    }

    static AuthorisationViewState loading() {
        return LOADING_STATE;
    }

    static AuthorisationViewState error(Throwable t) {
        return new AuthorisationViewState(Status.ERROR, t);
    }

}

File: Android/app/src/main/java/ru/inwords/inwords/authorisation/presentation/AuthorisationViewState.kt
package ru.inwords.inwords.authorisation.presentation

data class AuthorisationViewState(
        val status: Status,
        val errorMessage: String? = null
) {
    enum class Status {
        LOADING, SUCCESS, ERROR
    }

    companion object {
        private val SUCCESS_STATE = AuthorisationViewState(Status.SUCCESS, null)
        private val LOADING_STATE = AuthorisationViewState(Status.LOADING, null)

        internal fun success(): AuthorisationViewState {
            return SUCCESS_STATE
        }

        internal fun loading(): AuthorisationViewState {
            return LOADING_STATE
        }

        internal fun error(errorMessage: String): AuthorisationViewState {
            return AuthorisationViewState(Status.ERROR, errorMessage)
        }
    }
}
